libsolidity/ast/AST.cpp

```diff
@@ -34,18 +34,36 @@ using namespace std;
 using namespace dev;
 using namespace dev::solidity;
 
+class IDDispenser
+{
+public:
+	static size_t next() { return ++instance(); }
+	static void reset() { instance() = 0; }
```

Comment:
Currently left-over ASTnodes' IDs might be used again.

I think it makes sense to add a counter called `generation` and increment the generation every time `reset()` is called.

When an ID of an AST node of a previous generation is wanted, we can raise an error.